File: backend/app/services/guardrails.py
"""Citation integrity checks and hallucination guardrails.

Post-processes LLM agent outputs to:
  1. Validate that cited document IDs exist in context
  2. Check that quoted text actually appears in source chunks
  3. Flag low-confidence or unsupported claims
  4. Strip fabricated citations
"""
from __future__ import annotations

import logging
from difflib import SequenceMatcher

logger = logging.getLogger(__name__)


def validate_citations(answer: dict, context_chunks: list[dict]) -> dict:
    """Validate and filter citations in a QA answer against source chunks.

    Returns the answer dict with:
      - ``citations`` stripped of any that can't be traced back
      - ``_guardrail`` metadata added
    """
    raw_citations = answer.get("citations") or []
    if not raw_citations:
        answer["_guardrail"] = {"checked": True, "removed": 0, "kept": 0}
        return answer

    # Build lookup sets from the context we actually sent to the LLM
    known_doc_ids = {c.get("document_id") for c in context_chunks if c.get("document_id")}
    known_titles = {c.get("document_title", "").lower() for c in context_chunks if c.get("document_title")}
    chunk_texts = [c.get("text", "") for c in context_chunks]

    valid = []
    removed = 0
    for cit in raw_citations:
        # 1. Check document_id or title exists in context
        cid = cit.get("document_id", "")
        ctitle = (cit.get("document_title") or "").lower()
        id_ok = cid in known_doc_ids if cid else False
        title_ok = any(ctitle and ctitle in kt for kt in known_titles) if ctitle else False

        if not id_ok and not title_ok:
            logger.info("Guardrail: removed citation with unknown source %s / %s", cid, ctitle)
            removed += 1
            continue

        # 2. If chunk_text provided, verify it roughly matches something in context
        quoted = cit.get("chunk_text", "")
        if quoted and len(quoted) > 20:
            best_ratio = max(
                _fuzzy_contains(chunk, quoted) for chunk in chunk_texts
            ) if chunk_texts else 0.0
            if best_ratio < 0.4:
                logger.info("Guardrail: removed citation with unverifiable quote (best=%.2f)", best_ratio)
                removed += 1
                continue

        valid.append(cit)

    answer["citations"] = valid
    answer["_guardrail"] = {"checked": True, "removed": removed, "kept": len(valid)}
    return answer


def check_confidence(answer: dict) -> dict:
    """Downgrade confidence if the answer text contains hedging language
    or the model itself signalled uncertainty."""
    text = (answer.get("answer") or "").lower()
    confidence = answer.get("confidence", "medium")

    hedging = [
        "i'm not sure", "i am not sure", "it is unclear",
        "the context does not", "no relevant information",
        "cannot determine", "not enough information",
        "i don't have", "i do not have",
    ]
    if any(h in text for h in hedging):
        confidence = "low"

    answer["confidence"] = confidence
    return answer


def apply_guardrails(answer: dict, context_chunks: list[dict]) -> dict:
    """Run all guardrail checks on a QA answer."""
    answer = validate_citations(answer, context_chunks)
    answer = check_confidence(answer)
    return answer


# ── helpers ──────────────────────────────────────────────────────────────

def _fuzzy_contains(haystack: str, needle: str) -> float:
    """Return a 0-1 score for how well *needle* is contained in *haystack*."""
    if not needle or not haystack:
        return 0.0
    # Quick exact substring check
    if needle.lower() in haystack.lower():
        return 1.0
    # Fall back to fuzzy ratio on overlapping windows
    needle_lower = needle.lower()[:200]
    haystack_lower = haystack.lower()
    return SequenceMatcher(None, needle_lower, haystack_lower).ratio()

File: lib/guardrails.ts
// lib/guardrails.ts
// Citation validation & hallucination checks. Ported from ResNex-AI guardrails.py

export interface Citation {
  document_id?: string
  document_title?: string
  chunk_text?: string
  page_number?: number
  similarity?: number
  [key: string]: unknown
}

export interface GuardrailMeta {
  checked: boolean
  removed: number
  kept: number
}

export interface QAAnswer {
  answer: string
  citations: Citation[]
  confidence: 'high' | 'medium' | 'low'
  _guardrail?: GuardrailMeta
  [key: string]: unknown
}

export interface ContextChunk {
  document_id?: string
  fileName?: string
  content?: string
  text?: string
  [key: string]: unknown
}

function normalizeRef(text: string): string {
  return text
    .toLowerCase()
    .replace(/\.[a-z0-9]+$/i, '')
    .replace(/[_-]+\d+$/g, '')
    .replace(/[^a-z0-9]+/g, ' ')
    .trim()
}

/** Simple fuzzy containment — returns 0-1 score */
function fuzzyContains(haystack: string, needle: string): number {
  if (!needle || !haystack) return 0
  const h = haystack.toLowerCase()
  const n = needle.toLowerCase().slice(0, 200)
  if (h.includes(n)) return 1
  // Check if most words of needle appear in haystack
  const words = n.split(/\s+/).filter((w) => w.length > 3)
  if (words.length === 0) return 0
  const matched = words.filter((w) => h.includes(w)).length
  return matched / words.length
}

/** Remove citations that can't be traced back to any context chunk */
export function validateCitations(answer: QAAnswer, contextChunks: ContextChunk[]): QAAnswer {
  const rawCitations = answer.citations || []
  if (rawCitations.length === 0) {
    answer._guardrail = { checked: true, removed: 0, kept: 0 }
    return answer
  }

  const knownIds = new Set(contextChunks.map((c) => c.document_id).filter(Boolean))
  const knownTitles = contextChunks.map((c) => ({
    raw: c.fileName || '',
    normalized: normalizeRef(c.fileName || ''),
  })).filter((c) => c.raw)
  const chunkTexts = contextChunks.map((c) => c.content || c.text || '')

  const valid: Citation[] = []
  let removed = 0

  for (const cit of rawCitations) {
    const cid = cit.document_id || ''
    const ctitle = cit.document_title || ''
    const normalizedTitle = normalizeRef(ctitle)

    const idOk = cid ? knownIds.has(cid) : false
    const titleOk = normalizedTitle
      ? knownTitles.some(({ normalized }) =>
          normalized.includes(normalizedTitle) ||
          normalizedTitle.includes(normalized) ||
          fuzzyContains(normalized, normalizedTitle) >= 0.6
        )
      : false

    if (!idOk && !titleOk) {
      removed++
      continue
    }

    const quoted = cit.chunk_text || ''
    if (quoted && quoted.length > 20) {
      const bestRatio = chunkTexts.length
        ? Math.max(...chunkTexts.map((c) => fuzzyContains(c, quoted)))
        : 0
      if (bestRatio < 0.4) {
        removed++
        continue
      }
    }

    valid.push(cit)
  }

  answer.citations = valid
  answer._guardrail = { checked: true, removed, kept: valid.length }
  return answer
}

/** Downgrade confidence if answer contains hedging language */
export function checkConfidence(answer: QAAnswer): QAAnswer {
  const text = (answer.answer || '').toLowerCase()
  const hedging = [
    "i'm not sure", 'i am not sure', 'it is unclear',
    'the context does not', 'no relevant information',
    'cannot determine', 'not enough information',
    "i don't have", 'i do not have',
  ]
  if (hedging.some((h) => text.includes(h))) {
    answer.confidence = 'low'
  }
  return answer
}

/** Run all guardrail checks on a QA answer */
export function applyGuardrails(answer: QAAnswer, contextChunks: ContextChunk[]): QAAnswer {
  answer = validateCitations(answer, contextChunks)
  answer = checkConfidence(answer)
  return answer
}
